kotlin/internal/jvm/compile.bzl

```diff
@@ -428,39 +482,227 @@ def _run_kapt_builder_actions(
 
 def _run_ksp_builder_actions(
         ctx,
-        rule_kind,
         toolchains,
         srcs,
         compile_deps,
-        deps_artifacts,
-        annotation_processors,
-        transitive_runtime_jars,
-        plugins):
-    """Runs KSP using the KotlinBuilder tool
+        transitive_runtime_jars):
+    """Runs KSP2 via the KotlinBuilder with worker support.
 
     Returns:
         A struct containing KSP outputs
     """
     ksp_generated_java_srcjar = ctx.actions.declare_file(ctx.label.name + "-ksp-kt-gensrc.jar")
     ksp_generated_classes_jar = ctx.actions.declare_file(ctx.label.name + "-ksp-kt-genclasses.jar")
 
-    _run_kt_builder_action(
-        ctx = ctx,
-        rule_kind = rule_kind,
+    # Declare output directories
+    ksp_kotlin_output_dir = ctx.actions.declare_directory(ctx.label.name + "-ksp-kotlin")
```

Comment:
Most of this action should be done inside a worker, without the use of tree artifacts. The current implementation will be fairly slow locally (as it will need to compress and extract artifacts both via the zipper and via the tree artifact) and incredibly slow via remote execution.

Right now, if a source file changes it will retrigger 5 actions. By moving it to a worker that handles the staging of the files and any necessary decompressing, it will cut down to one.